rct229/rules/section5/section5rule21.py

```diff
@@ -0,0 +1,173 @@
+
+            return {
+                **data,
+                "total_fenestration_area_b": sum(
+                    find_all("$..total_window_area", window_wall_areas_dictionary_b),
+                    ZERO.AREA,
+                ),
+                "total_fenestration_area_p": sum(
+                    find_all("$..total_window_area", window_wall_areas_dictionary_p),
+                    ZERO.AREA,
+                ),
+                "surface_conditioning_category_dict_b": get_surface_conditioning_category_dict(
+                    data["climate_zone"], building_b
+                ),
+            }
+
+        def create_context_list(self, context, data=None):
+            building_b = context.baseline
+            building_p = context.proposed
+
+            scc = data["surface_conditioning_category_dict_b"]
+
+            baseline_surfaces = find_all("$..surfaces[*]", building_b)
+            proposed_surfaces = find_all("$..surfaces[*]", building_p)
+
+            matched_proposed_surfaces = match_lists_by_id(
+                baseline_surfaces, proposed_surfaces
+            )
+            proposed_baseline_surface_pairs = zip(
+                baseline_surfaces, matched_proposed_surfaces
+            )
+            # List of all baseline above grade wall to become the context for AboveGradeWall
+            return [
+                UserBaselineProposedVals(None, surface_b, surface_p)
+                for surface_b, surface_p in proposed_baseline_surface_pairs
+                if get_opaque_surface_type(surface_b) == OST.ABOVE_GRADE_WALL
+                and scc[surface_b["id"]] != SCC.UNREGULATED
+            ]
+
+        class AboveGradeWallRule(RuleDefinitionBase):
+            def __init__(self):
+                super(Section5Rule21.BuildingRule.AboveGradeWallRule, self).__init__(
+                    rmrs_used=UserBaselineProposedVals(False, True, True),
+                    required_fields={
+                        "$": ["construction"],
+                        "construction": ["u_factor"],
+                    },
+                )
+
+            def get_calc_vals(self, context, data=None):
+                above_grade_wall_b = context.baseline
+                above_grade_wall_p = context.proposed
+
+                def _helper_calc_val(above_grade_wall):
+                    """Helper function for calculating the total fenestration area for an above grade wall"""
+                    return sum(
+                        [
+                            subsurface.get("glazed_area", ZERO.AREA)
+                            + subsurface.get("opaque_area", ZERO.AREA)
+                            for subsurface in find_all(
+                                "subsurfaces[*]", above_grade_wall
+                            )
+                            if (
+                                getattr_(subsurface, "subsurface", "classification")
+                                == DOOR
+                                and (
```

Comment:
Will Black allow us to move the left paren on line 133 to line 131?  Otherwise  I'm not immediately sure of the precedence of something like   X and Y or Z.  Is it `(X and Y) or Z`,  or is it `X and (Y or Z)`? Moving the paren to line 131 would make our intent clearer in any case.

Reply:
It doesnt seem black will do that. I can add the parenthesis to the logic - indeed the logic is `(X and Y) or Z`. Adding parent will make the logic clearer